### ISSUE TITLE
FreqTrade Governance Policy 

### ISSUE THREAD

**Issue by mishaker** [freqtrade/freqtrade#1180]
## Introduction
This document is an attempt to improve FT community governance policy. it is based on my recent research on the history and the governance policies of few big and successful open source project. Almost none of below ideas is original. These are policies which are being used currently in open source projects. I just complied them and tried to adapt them to FT community.

## Background
Since few months ago there is an increasing frustration about how FT contribution process is functioning. we start seeing people leaving the project or at best being reluctant to participate or contribute to the project.
There is a good and a bad news about it. 
- The **good news** is that almost every open source project is facing the same challenge when their user base is growing. so it is not specific to us. 
- The **bad news** is that at this stage if we don't react and work on a clear governance policy then almost certainly in a near future the community will be dead and buried. 

I am going to propose a clear governance policy. needless to say, all people are welcome to comment and elaborate. before going into details, I am going to define our vocabulary as well as a concept widely known as "healthy open source community". 

## Vocabulary
* A **User** is an individual who uses FT.
* A **Contributor** is any individual creating or commenting on an issue or pull request.
* A **Committer** is a subset of contributors who have been given write access to the repository.
* A **TC** (Technical Committee) is a group of committers representing the required technical expertise to resolve rare disputes.
* A **Member** can be a committer or/and a member of TC.

## Healthy open source community
At left you see how a healthy open source community should look like and at right an unhealthy one which leads to frustrations:

![image](https://user-images.githubusercontent.com/939309/44859097-bd2f2480-ac73-11e8-8c9e-f3e0650bd050.png)

You notice that as the demands on the project from increased users rise, so do the contributors, and as contributors increase more are converted into committers. As the committer base grows, more of them rise to the level of expertise where they should be involved in higher level decision making.
The community would become unhealthy if there is a significant gap between lets say number of active committers and number of active contributors over time.

## Contributions
Any change to resources in this repository must be through pull requests. This applies to all changes to documentation, code, binary files, etc. Even long term committers and TC members must use pull requests. 

- No pull request can be merged without being reviewed.
- A PR can be merged after being reviewed at least by 3 committers.
- In the case of an objection being raised in a pull request by another committer, all involved committers should seek to arrive at a consensus by way of addressing concerns being expressed by discussion, compromise on the proposed change, or withdrawal of the proposed change.
- If a contribution is controversial and committers cannot agree about how to get it to land or if it should land then it should be escalated to the TC.

The only exception to the rules above is **exp** branch (see FT Particularity section)

## Becoming a committer
All contributors who land a non-trivial contribution should be on-boarded in a timely manner, and added as a committer, and be given write access to the repository.

Committers are expected to follow this policy and continue to send pull requests, go through proper review, and have other committers merge their pull requests.

This is they key to scaling contribution growth: committer growth.

## TC Process

The TC uses a "consensus seeking" process for issues that are escalated to the TC. The group tries to find a resolution that has no open objections among TC members. If a consensus cannot be reached that has no objections then a majority wins vote is called. It is also expected that the majority of decisions made by the TC are via a consensus seeking process and that voting is only used as a last-resort.

Resolution may involve returning the issue to committers with suggestions on how to move forward towards a consensus. It is not expected that TC will resolve all issues on its agenda and may prefer to continue the discussion happening among the committers.

Members can be added to the TC at any time. Any committer can nominate another committer to the TC and the TC uses its standard consensus seeking process to evaluate whether or not to add this new member. 

## Member resignation 
Members who do not participate consistently at the level of a majority of the other members are expected to resign.

## FT Particularity
Although technically robust and well designed, FT in its current state does not reflect some trading realities. this leads to serious difficulties in defining a development roadmap. many contributors want to collaborate on interdependence features and big refactoring but they have no collaboration space. 
One solution is to have a branch called **exp** on which committers can push the code without following contribution rules. 

- **exp** branch goal is to provide a quick collaboration tool in order to define the global FT development roadmap. it is by no mean supposed to be used in any environment. 
- **exp** branch is dropped and re-branched from **dev** every two months unless TC decides to extend the duration.
- There should be a clear explanation on README of the branch that it is temporary and it is used to define the development roadmap.
- Once the global development roadmap is defined the **exp** will be dropped permanently.

**Comment by xmatthias** [comment]
Hi there,

first of all, thanks for this.

I have a few comments to make.
> A PR can be merged after being reviewed at least by 3 committers.

The purpose of a review is to have more eyes on code, for small changes, 1 additional pair should be enough, for bigger changes more eyes can be needed. 
The "more code, more reviews" principle is applied at the moment, too - but can lead to longer lead-times.

Everyone (User, Commiter, ...) can (and honestly, should) do a review. Every Review is important and must be considered, so everyone can contribute. Also, there is a lot to learn by doing intensive code reviews - but this is not the place to discuss what that is.

This is true right now as well as it will be in the future,
#### however 
It's not what is happening currently, but I don't think it's the governance model which is at fault, but the community which somehow mostly expects only core-devs to participate in reviews, which does increase time-pressure on core-team members, and leads to bad community interactions, as everyone wants their code to be reviewd "right now".


Some PR's currently have a long time between opening and merging. 
I see 2 reasons for this:

* They are WIP - so the open -> close looks longer than it really is (#1101 - which had the [WIP] removed only 6 days ago).
* Missing reviews - community expects core to review (sample #1101, #1119 - everyone here can review this - apart from the people who contributed the change).

> All contributors who land a non-trivial contribution should be on-boarded in a timely manner, and added as a committer, and be given write access to the repository.

I only half agree with this point. One non-trivial contribution should not be enough to get onboarded, but continuous contributions and community involvement should be needed for this to happen. Honestly, this is happening too - that's how I got in on this project earlier this year. But it was not one contribution, but continuous involvement over a few months.

95% of the tasks team-members do can be done by everyone. No write-access is necessary for the following tasks:
* everyone can help a user in an issue
* everyone can review and approve / request changes on a PR
* everyone can open a PR

Write-access is only needed to "click the button" and merge, or close an issue, which (honestly) - is the easiest of all of the above things, if the issue has been answered - a good Review has been done already by someone else.

Personally, I think the focus on write-access is based on wrong assumptions, and an expectation to get something it isn't, which can lead to even more frustration.


I have a Question for you - as we've talked on slack about this too:
What would you expect to change from gaining write-access?


### Exp-branch

Dropping the branch every 2 months will lead to a lot of frustration in my opinion, as the changes people contributed and did not save will simply be dropped, unless someone does good lobbying with the TC to keep the branch for longer. (keep one thing about this community in mind: Not everyone is a developer/git wizzard, knowing how to do a feature branch for yourself to eventually add your changes to the new exp branch, but merge it all into one branch before pushing, ...).
Even a small improvement could help a lot - but with this approach it's **very** dangerous that this will simply be deleted.

To be honest, we currently have this already with `feat/backslap` and `feat/strategy` - where we defined people responsible for each branch - which should do the review, and core only pushes the "merge" button, which - as mentioned above - is the smallest of all steps.

#1147 shows that this did get no involvement, as no-one feels responsible to even look at the proposed refactoring.
Having everyone simply write to it will lead to chaos - one person wants it another way and overwrites the changes from the day before ... - frustration and a lot of friction will be the result of this i think.

#### Nonetheless - one very important question about this:
* If the exp branch leads to a good and valid improvement which should be added to the "regular branches" - who is responsible to properly develop / complete this feature, to write tests for it, integrate it to develop, assuming the initial developer of this improvement does not care/have time to finish it - or lacks skills in writing tests?


## wrapping up
This got longer than i wanted it to, sorry for this - but take it as a sign that i do take your considerations seriously...

Overall, i think we need to get more community involvement, not just in PR's, but also in Reviewing / helping other users with issues, etc.

However, I'm not certain that a governance model is what is missing. In this community, most people want to make money (that's what a trading bot should do, so it's understandable) - reviewing a PR does not produce money if you can wait for someone else 
 (core?) to review the PR - and that's what I feel is currently happening for the most part.

**Comment by mishaker** [comment]
@xmatthias First of all thank you for your exhaustive response. before going further and discuss about your points I would like to emphasize again that I hardly invented anything. the proposition is based on the governance policy and experience of Node.JS and Rust. 

I was always amazed by the speed of development of those two projects and now I understand better why they grew that fast. I also have to mention that I studied the model of some other projects but Node.JS and Rust seemed to be more pertinent.  

Most of the sentences above are just a copy paste of Node.js contribution policy which you can read for yourself here: https://github.com/nodejs/TSC/blob/master/BasePolicies/CONTRIBUTING.md

Before delving into technical details, I think there are two fundamental points to consider:

- Phrases like "everyone can review.." or "95% tasks can be done by ..." or "everybody should ..." are technically correct but it simply ignores the incentive people should have in order to do it. there is no clear reward/promotion or "what happens next?" explanation in the community reducing the contribution incentive. people should be rewarded by a way or another of their contribution. contributors should have the feeling of ownership and decision vote on the project or at least they should know how to gain it. this policy on FT is simply non-existent. notice that I am not defending my proposition here, I am just pointing out that reward/promotion policy in FT is non-existent. 

- "Having everyone simply write to it will lead to chaos": I assume by "everyone" you mean committers. committers are not everyone. they have already contributed non-trivial features to the project and they were promoted by other committers. after all, the reality is that people do know their limitations and want to defer hard decisions to people they know have more experience. according to Node.js community in rare case of conflicts between committers the decision is escalated to TC. Hence IMHO, your fear of chaos is exaggerated.

Lets jump into technical concerns one by one:

> A PR can be merged after being reviewed at least by 3 committers.

Node.js has different rules for "trivial" and "non-trivial" contributions. I think the definition is quite obvious and self-explanatory. a typo correction is trivial / CCXT integration is non-trivial. this is how Node.js handles non-trivial PRs: 

>_For non-trivial contributions, pull requests should sit for at least 36 hours to ensure that contributors in other timezones have time to review. Consideration should also be given to weekends and other holiday periods to ensure active committers all have reasonable time to become involved in the discussion and review process if they wish._

On top of that rule if we add "at least X committers should review" and "there should be no objection from any other committer" then we should be fine. 

>All contributors who land a non-trivial contribution should be on-boarded in a timely manner, and added as a committer, and be given write access to the repository => I only half agree with this point. One non-trivial contribution should not be enough ...

Then put it to two non-trivial contributions or propose another thing. the point is to give a clear vision so that contributor knows how he/she is rewarded.

>I have a Question for you - as we've talked on slack about this too:
What would you expect to change from gaining write-access?

More fluid and more responsible and more agile community. Let me be clear here: I did not find a single successful open source project in which write-access is limited to a group of 3 or 4 people who are mostly absent during long period of time. again I am going to cite Node.js policy here: Members who do not participate consistently at the level of a majority of the other members are expected to resign. 
If I am an active committer now and next month I know I will be busy for 2 months and cannot work at the pace of other committers then I will simply resign and let know others. if then after 3 months I am back in the hood I will regain access.
If your concern is that a committer who has a write access can do something wrong:
- Committer became committer because he proved his engagement to the community so chances he will mess up with his reputation is very low.
- At any moment TC can remove write-access from any committer. 

>Dropping the branch every 2 months will lead to a lot of frustration in my opinion

This was my idea to tackle the concern of cores about **exp** being used instead of **dev** over time. The best way is to let TC decide. but if you have another idea I will be happy to follow.

>If the exp branch leads to a good and valid improvement which should be added to the "regular branches" - who is responsible ...

You are looking at **exp** too technical. it isn't ! and it won't be. it is simply a working space. call it a jungle. "branch" nomenclature is not suitable for it. it is to provide a quick way to draft some concepts by more than one committer to feed the roadmap of development. the real development will happen as usual by PRs. 
I give a concrete example: I am developing money management now but for that I really need two bricks: 1) Async 2) Backslap. all I want to do is to draft money management to show to other members here in order to see if trading wise and technical wise I am on the right track. notice that at this level I don't even care if there are bugs in Async or Backslap, it is not my concern. I can simply merge those two branches into **exp** draft my money management feature, then showing to it to members. once I have enough feedback then I will formalize it in an issue then I will do a PR. the advantage of it is that at the end of the day those issues documented based on **exp** can be used to define the roadmap of development. once the roadmap is defined the project will be accelerated considerably as we give a clear vision to new contributors.

>However, I'm not certain that a governance model is what is missing ...

At this point and based on what I saw on other projects I can tell you with a 100% certainty that, like it or not, the governance model is non-existent in FT. either we should work on it or letting the community gradually disappear.

**Comment by mishaker** [comment]
> But if another of the selected commiters works on something else your work breaks ... ?
that's a conflict point - yes, different exp-branches can fix that - but who will cleanup exp1, exp2, exp3, ... .

Then let them fight to death ! I said it would be a jungle. if you go there then you assume your responsibility. we don't have to control everything. as long as we say clearly that it is an experimental broken branch ... . after all, you are in control, look at it as an experimentation. if it doesn't work then there is a clear sentence in Code Of Conduct saying that TC can drop the branch whenever needed. please don't look at it that technically. it is something wanted by the community and it doesn't cost that much. think about it in terms of opportunity and costs. At worst it doesn't lead to anything but it revives the dynamism, makes so many interesting discutions publicly in FT, it attracts other contributors, etc. At best it is going to be a liberal collaboration space. one which is not controlled by anyone, and the result of it would be a great roadmap for FT.  

>Honestly, the process on gaining access as you describe it has more or less been followed by the others to add me - but there is 0 documentation on how and why, so noone really knows about it.

Ok, then document it. try to formalize it. as an outsider if I start contributing to this project, it would be really motivating for me to know when I can be an insider. when can I feel that this project is mine then doubling my efforts. it is just human. 

>Feel free to do a PR proposing changes to contributing.md ...

I am sorry but I am not going to spend another minute on this project. it is been months that we are discussing about governance problems on slack then you told me we should document an issue then I put 3 days researching about governance of open source projects. it is important to mention that during all that time we didn't receive a single feedback from other owners of this project. this means a clear "je m'en foutisme" attitude. It doesn't change anything If I do a PR. If we cannot change their attitude that means we are not in a right community.

> How to show the roadmap? how to vote openly on what to bring to the roadmap?

We don't know it yet and we don't have to know it. most people in this community are seniors. I trust they can figure it out. we have to stop being a control freak.

>Comparing freqtrade to node.js is a dangerous comparison ..

Basically what you are saying is that a project with 2000+ contributors can work and emerge with a liberal governance policy but we as a 41+ contributors (most of us seniors) cannot put more liberal policies in our community? is that your argument ?

Moreover, Node.js started to thrive when they put those liberal governance policies. they were not 2000+ contributors when they started but they gained those contributors thanks to their liberal policy. 

PS: I appreciate the time you put on answering and elaborate on the project this weekend. please don't take personally any of above sentences. my language may be a little harsh but all I am trying to do is to revive this community.

**Comment by shusso** [comment]
Thanks for the suggestions @mishaker 👍 

>  this means a clear "je m'en foutisme" attitude

or then @xmatthias has been the one carrying the torch as the rest of us have had a bit too busy schedules. Which leads to another point. In my opinion are we trying to think of a "complex governance" schema on a problem, which to me looks like *"how do we move forward and add new features fast"*. Which is simple. Small PR's. Smaller the PR, faster it gets in. The faster smaller PR's get in, the faster we move and ship new features and cause friction to users.
*Assume* who ever you are sending the PR to, has 9-5 job, kids and hobbies. Then ask yourself *"if I only had 30mins weekend mornings to work on the bot and do PR reviews, would I be able to review and merge this"*.

> as an outsider if I start contributing to this project, it would be really motivating for me to know when I can be an insider. when can I feel that this project is mine then doubling my efforts

I think most of us, in the beginning, felt that this project was ours even when we were outsiders..

> revive this community

I salute you 🎩 I think at least for me, at some point I started feeling that "this is like work with project task trackers" and "PM's" yelling ship faster, but without actually getting paid. It kinda made it feel "not fun". And then when someone throws a mega PR and complains that "my PR has been open for day" while there are no tests, or even worse, the tests are failing and/or there is an expectation that the core members should finish the "rest of it"..

**Comment by mishaker** [comment]
Thanks @shusso for your response. would you mind If I ask you to read entirely the thread from the beginning please?

**Comment by xmatthias** [comment]
One resource i found about this topic is the following EuroPython talk: https://youtu.be/3pokUifUyWM

I think it outlines the main theme of this issue quite nicely, especially on where we can improve.

**Comment by xmatthias** [comment]
resolving as #1206 is merged. feel free to reopen if you feel differently.
